FWCore/Framework/interface/global/implementors.h

```diff
@@ -53,7 +55,9 @@ namespace edm {
       private:
         void preallocStreams(unsigned int iNStreams) final {
           caches_.resize(iNStreams,static_cast<C*>(nullptr));
+          preallocStreamsExtend(iNStreams);
         }
+        virtual void preallocStreamsExtend(unsigned int iNStreams) {}
```

Comment:
what is the new `preallocStreamsExtend` method supposed to be used for ?

Reply:
Then I would suggest to drop `Extend` from the name, since this will likely be called only once at the beginning of the job, and I would favour a name like `preallocate` or `reserve` to match what we use for services or for vectors.

Is this only for `global` and `limited`, or also for `stream` and `one` modules ?
For a `limited` module, does it return the number of streams or the maximum concurrency of the module ?

Finally, this seems to give an alternative to the `StreamCache` approach: one can just have a `vector` as a data member, `resize()` it in the `preallocate` method, and use it with the `StreamID` in the `analyze`/`produce`/`filter` method, right ?

Reply:
I will change the name to preallocate. That is much better. (The reason we used extend was related to things internal to the Framework that a module writer should not need to know about ...). It is only for global and limited modules. The other kinds of modules should not need to know the number of streams. For both limited and global, it refers to the total number of streams not the maximum concurrency. Your third point is correct. If you look inside of StreamCacheHolder it just contains a vector of pointers to the cached objects in addition to some other things (mainly a larger interface of available functions). It is supposed to make it easier for a module writer not to have to implement the code related to the vector. I used preallocate in the test code because I wanted to construct a single object (the WaitingServer which handles all streams) and also wanted it initialized earlier than the point where the StreamCache is available.